docs: "Catering to the client-side" blog post (#265)
open-feature/docs.openfeature.dev: blog/06-02-2023-catering-to-the-client-side.md

---
slug: "catering-to-the-client-side"
title: "Catering to the client-side"
date: 2023-02-06
authors: [petehodgson]

> description: "TODO"

Reviewer: Don't forget to add a short description.

> description: "TODO"

Reviewer 2: ```suggestion
description: "TODO"
tags: [some, relevant, keywords]
```

draft: false
---

While OpenFeature initially focused on support for server-side feature flagging, we know that a lot of feature-flagging (likely the majority) happens on the client - mobile apps and frontend web apps. As such, we're currently finalizing [a proposal](https://github.com/open-feature/spec/issues/167) which extends the OpenFeature spec to support client-side use cases. 

In this post I'll summarize those changes, but to understand them in context we'll first talk about what makes client-side feature flagging different before diving into how that will impact the OpenFeature APIs.

## Context is King: why client-side flagging is different

### Static evaluation context

The biggest difference between the client- and server-side feature flagging is the relative frequency of evaluation context changes within those two paradigms. In the typical server-side scenario - a service responding to http requests - the context for a feature flagging decision changes with every request. Every new request is coming from a new user, and a lot of the evaluation context which affects a feature flagging decision is based on the user making the request.

> In contrast, with a client-side app all feature flagging decisions are made in the context of the same user - the user interacting with the client-side app - and so the evaluation context is quite static. There are cases where evaluation context will change within a client-side app - when a user logs in, for example - but by and large with client-side code we can treat feature flag evaluation context as something that is fixed (while still providing mechanisms to update it).

Reviewer 2: ```suggestion
In contrast, with a client-side app all feature flagging decisions are made in the context of the same user - the user interacting with the client-side app - and so the evaluation context is relatively static. There are cases where evaluation context will change within a client-side app - when a user logs in, for example - but by and large with client-side code we can treat feature flag evaluation context as something that is fixed (while still providing mechanisms to update it).
```

Reviewer 2: Doesn't have to be this wording, but I'd prefer a comparative here instead of a superlative.

### Flag evaluation is slow

With server-side flags, we can assume that evaluating a feature flag is a relatively fast operation. In some scenarios the flagging rules can be evaluated in-process, and if not then flag evaluation is typically a quick service call - akin to making a DB query or calling a redis cache.

> This situation is quite different with client-side flags. It's typically not possible to evaluate a flagging decision locally. Local evaluation would require the full feature flagging ruleset to be exposed to the client, something many organizations are not comfortable with. Additionally, flag evaluation may require additional context that's simply not available on the client at the time of evalution. 

Reviewer 2: > Local evaluation would require the full feature flagging ruleset to be exposed to the client, something many organizations are not comfortable with.

This is the way some vendors do it though (I think 2 we know of). I think it's valid, though there are tradeoffs. One advantage of doing things this way is little (or possible no) PII leaves the client app for evaluation purposes. As we've mentioned before, SDKs built this way can conform to our proposed paradigm, while the reverse isn't true.

I'm not sure how valuable it is to mention this, but I would at the least, suggest that both patterns are valid.

Author: I agree that this model exists, but AFAICT it's pretty rare. Given that, I'd rather not distract the reader.

> This means that a client-side flag evalution requires a backend call. Unfortunately we should also anticipate the latency for such a call to be slow, particularly if our users are behind a spotty internet connection. In fact, with a native mobile app we have to handle a fully disconnected client.

Reviewer 2: ```suggestion
This means that a client-side flag evaluation frequently requires a backend call. Unfortunately we should also anticipate the latency for such a call to be slow, particularly if our users are behind a spotty internet connection. In fact, with a native mobile app we have to handle a fully disconnected client.
```

Author: I think it's a pretty rare case where it wouldn't require a backend call, and I'd like to avoid superfluous "caveating words" like "often", "in general", "frequently", etc.

### Pre-evaluation

> So client-side flag evaluation is slow, but we've also seen that the inputs into that flag evaluation - the evaluation context - are fairly static for client-side apps, and that means the *results* of flag evalution are fairly static too.

Reviewer 2: ```suggestion
So client-side flag evaluation is slow, but we've also seen that the inputs into that flag evaluation - the evaluation context - are fairly static for client-side apps, and that means the *results* of flag evaluation are fairly static too.
```

> How do we handle an expensive operation with fairly static results? We add caching! And that's what most client-side feature flagging frameworks do. Specifically, they do an optimistic pre-evaluation of all the feature flagging decisions that might be needed and then cache those decisions. Then whenever client-side code needs to make a flagging decision the framework simply returns the pre-evaluated result from its local cache. 

Reviewer 2: Well-put.

```mermaid
sequenceDiagram
  actor app
  participant client as feature flagging logic
  participant cache as in-memory cache
  participant provider as flag service
  app->>+client: userLoggedIn(userId)
  client->>+provider: evaluateFlags(evaluationContext)
  provider-->>-client: flag values
  client->>+cache: store(flag values)
  cache->>-client: .
  %% deactivate cache
  client->>-app: .
  %% deactivate client
  note right of app: some time later..
  app->>+client: operation that needs a flagging decision
  client->>+cache: getFlagValue(flagKey)
  note right of cache: no call to flagging service needed
  cache->>-client: previously evaluated value
  client->>-app: .

```

> Put another way, we seperate flag [**evaluation**](/docs/specification/glossary#evaluating-flag-values) - passing an evaluation context through a set of rules in order to determine a flagging decision - from flag [**resolution**](/docs/specification/glossary#resolving-flag-values) - getting the flagging decision for a specific feature flag.

Reviewer 2: ```suggestion
Put another way, we separate flag [**evaluation**](/docs/specification/glossary#evaluating-flag-values) - passing an evaluation context through a set of rules in order to determine a flagging decision - from flag [**resolution**](/docs/specification/glossary#resolving-flag-values) - getting the flagging decision for a specific feature flag.
```

## Client-side support in OpenFeature

With OpenFeature we have been thinking about how to support these key differences between client-side and server-side feature flagging. We think about these differences as two paradigms: dynamic context (for server-side flags) and static context (for client-side flags). 

OpenFeature's current Evaluation API supports the dynamic paradigm quite nicely, but to support the static paradigm (and thus client-side flagging) we need to add a second flavor of the Evaluation API. 

### Server-side evaluation today
Let's compare and contrast. A typical server-side flagging decision using OpenFeature's current SDK might look something like this:
``` java
@GetMapping("/hello")
public String getSalutation() {
    final Client client = openFeatureAPI.getClient();

>     final evalContext:EvaluationContext = evalContextForCurentRequest();

Reviewer 2: ```suggestion
    final evalContext:EvaluationContext = evalContextForCurrentRequest();
```

    if (client.getBooleanValue("use-formal-salutation", false, evalContext)) {
      return "Good day to you!";
    }else{
      return "Hey, what's up?";
    }
} 
```
You can see that we're passing evaluation context every time we need to make a flagging decision.

### Client-side evaluation tomorrow
With the currently proposed OpenFeature changes, a client-side flagging decision would look more like this:

``` java
public string generateSalutation(){
  if (client.getBooleanValue("use-formal-salutation", false)) {
    return "Good day to you!";
  }else{
    return "Hey, what's up?";
  }
}
```

You can see we are no longer passing any evaluation context when requesting a flagging decisions. 

> But our flagging framework needs to get flag evaluation context from somewhere. So rather than passing the context as part of each flagging decision, the app provides a new evaluation context to OpenFeature whenever a user logs in or out (as part of this app's hypothetical `onAuthenticated(...)` handler): 

Reviewer: This makes it sound like the only place that context can change is during log in or out. Could you try and make it clear that this is just an example of one scenario where context may be updated?

Author: I gave this a try, but couldn't figure out a way to make this clear without distracting the reader from the main point - how an app might set evaluation context as a side effect of some context change within the app

Author: I gave this a try, but couldn't figure out a way to make this clear without distracting the reader from the main point - how an app might set evaluation context as a side effect of some context change within the app

``` java
// called whenever a user logs in (or out)
public void onAuthenticated(userId:String){
  OpenFeatureAPI api = OpenFeatureAPI.getInstance();

>   api.setEvaluationContext(new MutableContext().add("targettingKey", userId));

Reviewer: Perhaps we should use JavaScript in the example since it may be easier for a front-end dev to provide feedback. This example is pretty straightforward though, so it's completely up to you.

Author: I was reluctant to use JS for this first example because the reader might get distracted by the sync vs async difference in the two flavors of JS APIs. I also sort of think it would be nice to have something other than JS, for readers who are more Java than JavaScript.

Reviewer: Yep, that makes sense.

>   api.setEvaluationContext(new MutableContext().add("targettingKey", userId));

Reviewer 2: ```suggestion
  api.setEvaluationContext(new MutableContext().add("targetingKey", userId));
```

}
```

> This call to update the evalution context can prompt the configured flagging provider to update any cached feature flag values using the new evaluation context. The provider will be notified that the evaluation context has changed via a new `onContextSet` handler which is being added to the OpenFeature provider interface:

Reviewer 2: ```suggestion
This call to update the evaluation context can prompt the configured flagging provider to update any cached feature flag values using the new evaluation context. The provider will be notified that the evaluation context has changed via a new `onContextSet` handler which is being added to the OpenFeature provider interface:
```
